### PR TITLE
epoll_wait timeout reset when called multiple times

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
epoll_wait accepts a timeout argument which will specify the maximum amount of time the epoll_wait will wait for an event to occur. If the epoll_wait method returns for any reason that is not fatal (e.g. EINTR) the original timeout value is re-used. This does not honor the timeout interface contract and can lead to unbounded time in epoll_wait.

Modifications:
- The time taken by epoll_wait should be decremented before calling epoll_wait again, and if the remaining time is exhausted we should return 0 according to the epoll_wait interface docs http://man7.org/linux/man-pages/man2/epoll_wait.2.html

Result:
epoll_wait will wait for at most timeout ms according to the epoll_wait interface contract.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@normanmaurer - PTAL

**Comment by normanmaurer** [comment]
@Scottmitch this failed with:

```
[23:35:47][io.netty:netty-transport-native-epoll] /opt/jdk/1.7.0_80/jre/bin/java: symbol lookup error: /tmp/libnetty-transport-native-epoll6816250309117362204.so: undefined symbol: clock_gettime
```

Please check

**Comment by Scottmitch** [comment]
@normanmaurer - had to explicitly link `librt` for RHEL apparently. PTAL

File: transport-native-epoll/src/main/c/netty_epoll_native.c [hunk 1]
```diff
@@ -103,7 +104,49 @@ jfieldID packetPortFieldId = NULL;
...
+    return JNI_TRUE;
+  }
+  #endif
+  #ifdef CLOCK_REALTIME
+  epollWaitClock = CLOCK_REALTIME;
+  if (clock_gettime(epollWaitClock, &ts) != EINVAL) {
+    return JNI_TRUE;
+  }
+  #endif
+  fprintf(stderr, "FATAL: could not find a clock for clock_gettime!\n");
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/c/netty_epoll_native.c]: should we set epollWaitClock back to 0 ?

**Comment by normanmaurer** [comment]
@Scottmitch just one comment... also please squash.

File: transport-native-epoll/src/main/c/netty_epoll_native.c (hunk 1 above)
Comment (Scottmitch) [inline transport-native-epoll/src/main/c/netty_epoll_native.c]: 0 may not be a legitimate value. It turns out to be defined as `CLOCK_REALTIME` on my system (which should be implemented everywhere according to [1]) so I think if we get to this point we are in a pretty bad state and should fail loudly.

[1] http://linux.die.net/man/3/clock_gettime
> All implementations support the system-wide realtime clock, which is identified by CLOCK_REALTIME

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (52bbfd3) 4.0 (ed7c05f)

**Pull request closed without merging**